src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,304 @@
+            return JOINING_RESYNCING
+        elif state_check(oneOrMore=[NORMAL, WAITING_TO_LEAVE]):
+            return LEAVE_PENDING
+        elif state_check(oneOrMore=[NORMAL, LEAVING],
+                         zeroOrMore=[NORMAL_ACKNOWLEDGED_CHANGE,
+                                     LEAVING_ACKNOWLEDGED_CHANGE]):
+            return STARTED_LEAVING
+        elif state_check(oneOrMore=[NORMAL_ACKNOWLEDGED_CHANGE,
+                                    LEAVING_ACKNOWLEDGED_CHANGE],
+                         zeroOrMore=[NORMAL_CONFIG_CHANGED,
+                                     LEAVING_CONFIG_CHANGED]):
+            return LEAVING_CONFIG_CHANGING
+        elif state_check(oneOrMore=[NORMAL_CONFIG_CHANGED,
+                                    LEAVING_CONFIG_CHANGED],
+                         zeroOrMore=[NORMAL, FINISHED]):
+            return LEAVING_RESYNCING
+        elif state_check(oneOrMore=[NORMAL, FINISHED]):
+            return FINISHED_LEAVING
+        else:
+            # Cluster in unexpected state.
+            return INVALID_CLUSTER_STATE
+
+    # Returns the local node's state in the cluster, and None if the local node
+    # is not in the cluster.
+    def calculate_local_state(self, cluster_view):
+        return cluster_view.get(self._ip)
+
+    # Read the state of the cluster from etcd.
+    def read_from_etcd(self):
+        cluster_view = {}
+
+        try:
+            result = self._client.get(self._key)
+            cluster_view = self.parse_cluster_view(result.value)
+
+            # If the cluster view hasn't changed since we last saw it, then
+            # wait for it to change before doing anything else.
+            _log.info("Read cluster view {} from etcd, "
+                      "comparing to last cluster view {}".format(
+                          cluster_view,
+                          self._last_cluster_view))
+            if cluster_view == self._last_cluster_view:
+                while not self._terminate_flag:
+                    try:
+                        result = self._client.watch(self._key,
+                                                    index=result.modifiedIndex+1,
+                                                    timeout=5,
+                                                    recursive=False)
+                        break
+                    except urllib3.exceptions.TimeoutError:
+                        # Timeouts after 5 seconds are expected - just check if
+                        # we're termninating and continue
+                        pass
+                    except ValueError:
+                        # The index isn't valid to watch on, probably because
+                        # there has been a snapshot between the get and the
+                        # watch. Just start the read again.
+                        _log.info("etcd index {} is invalid, retrying".format(
+                            result.modifiedIndex+1))
```

Comment:
Is there any other reason this exception could be thrown? I'm worried about this tight looping